### PR TITLE
Improve early data examples

### PR DESCRIPTION
**Pull request by tahmid-23**
Adds simple_0rtt_server, lets simple_0rtt_client connect to arbitrary servers (so it can work with simple_0rtt_server), makes tlsserver-mio read early data

questions
1. I need to add `--package rustls-examples` or else I get an error about no crypto provider being installed for both the existing 0rtt client and the new 0rtt server. might I be doing something wrong?
2. should I use an actual named args struct like in tlsclient-mio for the new options, or are the positional arguments fine?
3. if not, do I have the correct "syntax" for the arguments in the usage comment?
4. the handshaking code is a bit obtuse and mostly inspired by complete_io(). ideally I want to be able to progress the handshake without completion (since the point of early data is to not have to wait for the handshake completion). should some of this code be extracted to a new method? or maybe we can have a new function to handshake until early data can be read?
5. afaik early data is terminated by the EndOfEarlyData message, not by handshake completion, but usually these happen at the same time. should there be a new method to determine this, or is it not really worth it?
6. tlsserver-mio early data does not "work" with `openssl s_client -connect localhost:8443 -tls1_3 -sess_in sess.pem -early_data early.txt` if I don't specify --package rustls-examples (at least on my machine). something to do with fips feature being enabled causing a HRR to be sent due to mismatched kx groups, meaning that the early data gets dropped (I barely know what any of that means but this is what I think is happening). is this anything to worry about?

### CODE REVIEW

**Comment by ctz** [comment]
> questions
> 
> 1. I need to add `--package rustls-examples` or else I get an error about no crypto provider being installed for both the existing 0rtt client and the new 0rtt server. might I be doing something wrong?

This is unfortunately due to feature unification in the workspace, coming from `bogo/Cargo.toml` which turns on everything.

> 2. should I use an actual named args struct like in tlsclient-mio for the new options, or are the positional arguments fine?

I think this is fine, for the level of complexity here. But if we add any options or further arguments then it may be better to reach for a proper library.

> 4. the handshaking code is a bit obtuse and mostly inspired by complete_io(). ideally I want to be able to progress the handshake without completion (since the point of early data is to not have to wait for the handshake completion). should some of this code be extracted to a new method? or maybe we can have a new function to handshake until early data can be read?

I think this would be better motivated if we enabled `send_half_rtt_data` and wrote some data at the point of early data receipt. WDYT?

> 5. afaik early data is terminated by the EndOfEarlyData message, not by handshake completion, but usually these happen at the same time. should there be a new method to determine this, or is it not really worth it?

See point (1) on #2085

> 6. tlsserver-mio early data does not "work" with `openssl s_client -connect localhost:8443 -tls1_3 -sess_in sess.pem -early_data early.txt` if I don't specify --package rustls-examples (at least on my machine). something to do with fips feature being enabled causing a HRR to be sent due to mismatched kx groups, meaning that the early data gets dropped (I barely know what any of that means but this is what I think is happening). is this anything to worry about?

This is fine, it is because openssl prefers x25519 which is disabled in fips mode. A HRR is needed to complete such a handshake. The alternative is to add `-groups P-256` to the openssl command.

**Comment by tahmid-23** [comment]
> I think this would be better motivated if we enabled `send_half_rtt_data` and wrote some data at the point of early data receipt. WDYT?

I think there is still some merit in being able to read the client's early data without sending half-RTT data if you choose not to have it enabled. It makes sense to show `send_half_rtt_data` in the example, but ideally you could handshake until early data is ready without `send_half_rtt_data`.

If we don't want to have the manual handshake code, would `complete_io()` finish once the client hello has been received, or is there a better solution? How might this work internally?

For clarification, the documentation for `send_half_rtt_data` says it "comes at the cost of a subtle weakening of the normal handshake integrity guarantees that TLS provides". What exactly does this mean? Is there any risk to sending half-RTT data if you are already accepting 0-RTT data?

> See point (1) on #2085

Wouldn't that be a breaking change?

**Comment by ctz** [comment]
> If we don't want to have the manual handshake code, would `complete_io()` finish once the client hello has been received, or is there a better solution? How might this work internally?

I feel like `complete_io` is a function enabling simpler downstream programs, and making that take into account this subtle behaviour might make those programs more complex? At the more complex programs do not (or should not) use `complete_io`.

> For clarification, the documentation for `send_half_rtt_data` says it "comes at the cost of a subtle weakening of the normal handshake integrity guarantees that TLS provides". What exactly does this mean? Is there any risk to sending half-RTT data if you are already accepting 0-RTT data?

When a server receives the client's finished message and it is valid, it learns that the handshake was successful in both directions. If it sends data before then, it does so before the handshake integrity is knowable. (This whole explanation ignores client auth.)

File: examples/src/bin/simple_0rtt_server.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
+//! This is an example server that streams 0-RTT early data from the client.
```
Comment (cpu) [inline examples/src/bin/simple_0rtt_server.rs:1]: I think [`examples/README.md`](https://github.com/rustls/rustls/blob/main/examples/README.md) needs an update for this new example. 

Similarly, the [daily-tests.yml](https://github.com/rustls/rustls/blob/8c04dba680d19d203a7eda1951ad596f5fc2ae59/.github/workflows/daily-tests.yml#L59-L60) has a section that could use an update for this binary.

File: examples/src/bin/tlsserver-mio.rs [hunk 2]
```diff
@@ -251,6 +251,19 @@ impl OpenConnection {
...
+                early_data
+                    .read_to_end(&mut buf)
+                    .unwrap();
+
+                if !buf.is_empty() {
+                    debug!("early data read {:?}", buf.len());
+                    self.incoming_plaintext(&buf);
+                    return;
+                }
+            }
```
Comment (cpu) [inline examples/src/bin/tlsserver-mio.rs:265]: Can you break this change out into its own commit alongside a commit message motivating the change?

File: examples/src/bin/simple_0rtt_client.rs [hunk 3]
```diff
@@ -9,17 +11,19 @@
...
 use std::net::TcpStream;
+use std::str::FromStr;
 use std::sync::Arc;
+use std::{env, fs};
 
 use rustls::pki_types::ServerName;
 use rustls::RootCertStore;
 
-fn start_connection(config: &Arc<rustls::ClientConfig>, domain_name: &str) {
+fn start_connection(config: &Arc<rustls::ClientConfig>, domain_name: &str, port: u16) {
```
Comment (cpu) [inline examples/src/bin/simple_0rtt_client.rs:21]: The changes to `simple_0rtt_client.rs` seem OK to me, but would be best in their own independent commit.

**Review by cpu — COMMENTED** [review]
Some initial feedback, thanks!

File: examples/src/bin/simple_0rtt_server.rs (hunk 1 above)
Comment (tahmid-23) [reply examples/src/bin/simple_0rtt_server.rs:1]: this example doesn't terminate, it keeps accepting connections

Comment (cpu) [reply examples/src/bin/simple_0rtt_server.rs:1]: Good point, and it doesn't have a `--help` or similar. Fair to skip the CI update then :+1:

**Review by cpu — APPROVED** [review]

**Comment by cpu** [comment]
@djc @ctz Should I hold this for your reviews or is it acceptable to merge?

File: examples/src/bin/simple_0rtt_client.rs [hunk 4]
```diff
@@ -1,5 +1,7 @@
 //! This is an example client that uses rustls for TLS, and sends early 0-RTT data.
 //!
+//! Usage: cargo r --bin simple_0rtt_client --package rustls-examples [domain name] [port] [path/to/ca.cert]
```
Comment (ctz) [inline examples/src/bin/simple_0rtt_client.rs:3]: I have a feeling that a `--` might be needed here, after `--package rustls-examples`?

File: examples/src/bin/simple_0rtt_server.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+//!
+//! Usage: cargo r --bin simple_0rtt_server --package rustls-examples <path/to/cert.pem> <path/to/privatekey.pem>
```
Comment (ctz) [inline examples/src/bin/simple_0rtt_server.rs:3]: Ditto (maybe? I haven't checked.)

Comment (tahmid-23) [reply examples/src/bin/simple_0rtt_server.rs:3]: nope it's fine

**Review by ctz — APPROVED** [review]
